internal/provider/gce/environ_firewall.go

```diff
@@ -62,10 +68,24 @@ func firewallSpec(name, target string, sourceCIDRs []string, ports protocolPorts
 // If a rule matching a set of source ranges doesn't
 // already exist, it will be created - the name will be made unique
 // using a random suffix.
-
 func (env *environ) OpenPorts(ctx context.ProviderCallContext, target string, rules firewall.IngressRules) error {
-	err := env.openPorts(ctx, target, rules)
-	return google.HandleCredentialError(errors.Trace(err), ctx)
+	// Separate IPv4 and IPv6 rules to avoid mixed CIDRs in a single firewall, which is disallowed by GCE.
+	// Ensure default SourceCIDRs are set for any rule with empty SourceCIDRs before separating rules.
+	rules.EnsureSourceCIDRs()
+	ipv4Rules := rules.RemoveCIDRsMatchingAddressType(network.IPv6Address)
+	if len(ipv4Rules) > 0 {
+		if err := env.openPorts(ctx, target, ipv4Rules); err != nil {
+			return google.HandleCredentialError(errors.Trace(err), ctx)
+		}
+	}
+
+	ipv6Rules := rules.RemoveCIDRsMatchingAddressType(network.IPv4Address)
+	if len(ipv6Rules) > 0 {
+		err := env.openPorts(ctx, target, ipv6Rules)
```

Comment:
This will just attempt to open a firewall with the same name `target` as was used for ipv4. If mixed cidrs are not allowed, then we'd need to use 2 differently named firewalls right?

Reply:
Yes, this will create two firewalls with different names, as we randomize the names based on the source CIDRs and the target for the mixed CIDRs case. 
https://github.com/juju/juju/blob/b10f458f15f4d4f57fa7ebce95c9d115608082c9/internal/provider/gce/environ_firewall.go#L156-L161
In such a case, the source CIDRs should be different, so in the end there will be two firewalls with different names, but using the same network target.